Call delete API in hover dropdown menu

joannag6 opened the pull request:
## Summary
When user clicks on delete in the hover menu, that media element is deleted from their media library. If successful, the element is removed visually from the library. If failed, the element stays and an error snackbar appears.

Displays a confirmation dialog informing the user that the deletion is permanent.

![Screen Shot 2020-05-29 at 3 19 11 PM](https://user-images.githubusercontent.com/15134432/83224030-c7bb7c80-a1bf-11ea-9035-a034ce0bac22.png)

## To-do
Turn on dropdown flag when Edit UI is merged.

joannag6 committed:
Call delete API in hover dropdown menu

barklund commented on assets/src/edit-story/components/library/panes/media/deleteDialog.js:
```diff
@@ -0,0 +1,132 @@
+  color: ${({ theme }) => theme.colors.fg.v0};
+`;
+
+const Bold = styled.span`
```
`styled.strong` seems more semantic.

joannag6 replied:
I just ended up using `<strong>` directly.

barklund commented on assets/src/edit-story/components/library/panes/media/deleteDialog.js:
```diff
@@ -0,0 +1,132 @@
+        });
+      }
+    };
+    shouldDelete ? deleteMediaItem() : null;
```
Why is it happening in an effect and not as a direct callback from the click instead?

joannag6 replied:
Moved it to a useCallback func that's directly called onClick. (Wasn't sure if async funcs could be called in onClick handlers.)

barklund commented on assets/src/edit-story/components/library/panes/media/deleteDialog.js:
```diff
@@ -0,0 +1,132 @@
+      <DialogBody>
+        {__(
+          'You are about to permanently delete this image/video from your site. ' +
+            'The image/video will appear broken in any stories that uses it. ',
```
Not just stories. also all other content pieces across the entire WP installation.

joannag6 replied:
I got this from the UX mocks, how about "any Wordpress content that uses it" ?

swissspidy replied:
Unresolving as it hasn't been adapted yet?

If not changing to "any WordPress content that uses it" (note the capital P), it should at least be "in any stories that **use** it." (note the missing `s`)

spacedmonkey replied:
How about using `sprintf` and doing something like this. 
```
You are about to permanently delete this %s from your site. Any references to this %s through your site will be broken
```

joannag6 replied:
Extracted them out to separate variables like in https://github.com/google/web-stories-wp/pull/1733/files/9c53338bb04e08fc00626f466969d30bdba87956#diff-f23faa359a2b930fadcc47d113718db5R101

barklund commented on assets/src/edit-story/components/library/panes/media/dropDownMenu.js:
```diff
@@ -85,32 +88,42 @@ function DropDownMenu({ pointerEntered, isMenuOpen, setIsMenuOpen }) {
+      {showDeleteDialog && (
+        <DeleteDialog
+          mediaId={mediaId}
+          showDeleteDialog={showDeleteDialog}
```
Isn't this just a boolean, that's always true?

barklund commented on assets/src/edit-story/components/library/panes/media/dropDownMenu.js:
```diff
@@ -85,32 +88,42 @@ function DropDownMenu({ pointerEntered, isMenuOpen, setIsMenuOpen }) {
+        <DeleteDialog
+          mediaId={mediaId}
+          showDeleteDialog={showDeleteDialog}
+          setShowDeleteDialog={setShowDeleteDialog}
```
Perhaps rename this prop (not the value) to `onClose` or similar more generic name.

joannag6 replied:
Done.

barklund commented on assets/src/edit-story/components/library/panes/media/deleteDialog.js:
```diff
@@ -0,0 +1,132 @@
+      actions={
+        <>
+          <Plain
+            data-testid="cancel"
```
Test id should not be necessary here - you can easily get this component by text or role. Test id should only be a last resort, when all else fails. And similarly below for `delete`.

joannag6 replied:
Hmm why is that? I'd think that `data-testid` is more a description of the functionality, which is something that is less likely to change than the text chosen.

barklund replied:
Unit tests are more useful ¹, the more they resemble real user behavior. Real users look for buttons by looking at their text, layout or assistive properties (like labels or roles) - not by looking for an element with a certain DOM data attribute.

See the [guiding principles](https://testing-library.com/docs/guiding-principles) behind testing-library, [their guide to choosing a query](https://testing-library.com/docs/guide-which-query) or the author, Kent C. Dodds', [own guides on the matter](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#using-the-wrong-query).

**In practical terms**: If the button suddenly changes text, users might be confused. It seems fair that your test would too.

Even better than a simple `getByText` (which might also find a headline or a paragraph) is `getByRole('button', { name: /cancel/i })` –  the ["accessible name"](https://www.w3.org/TR/accname-1.1/) here comes from the actual text on the button (with caveats such as `aria-` properties taking precedence if present) and would still break, if the text changes, but would not match other elements, that aren't buttons.

And note that you don't have to manually put a `role="button"` on the element - any `<button />` has the proper implicit role.

¹ For a well-chosen definition of "useful", that is 😉

joannag6 replied:
Ah thanks for the guides! Used `getByRole` :)

barklund reviewed (commented):
Various minor comments, overall looks good.

joannag6 committed:
Fixed PR comments
- renamed props
- used strong instead of bold
- use useCallback directly instead of useEffects
- renamed callback to onClose

joannag6 committed:
Mocked out providers in storybook stories

joannag6 committed:
Use getByRole instead of getByTestId

spacedmonkey commented on assets/src/edit-story/components/library/panes/media/deleteDialog.js:
```diff
@@ -0,0 +1,108 @@
+  return (
+    <Dialog
+      open={true}
+      onClose={() => onClose()}
```
```suggestion
      onClose={onClose}
```

spacedmonkey commented on assets/src/edit-story/components/library/panes/media/deleteDialog.js:
```diff
@@ -0,0 +1,108 @@
+      title={__('Delete image/video?', 'web-stories')}
+      actions={
+        <>
+          <Plain onClick={() => onClose()}>{__('Cancel', 'web-stories')}</Plain>
```
```suggestion
          <Plain onClick={onClose}>{__('Cancel', 'web-stories')}</Plain>
```

spacedmonkey commented on assets/src/edit-story/components/library/panes/media/deleteDialog.js:
```diff
@@ -0,0 +1,108 @@
+      actions={
+        <>
+          <Plain onClick={() => onClose()}>{__('Cancel', 'web-stories')}</Plain>
+          <Plain onClick={() => onDelete()}>
```
```suggestion
          <Plain onClick={onDelete}>
```

spacedmonkey commented on assets/src/edit-story/components/library/panes/media/deleteDialog.js:
```diff
@@ -0,0 +1,108 @@
+  } = useMedia();
+
+  const onDelete = useCallback(async () => {
+    onClose();
```
Should this be at the end of this async function?

joannag6 replied:
I put this here so it'll be harder for the user to accidentally try to delete the same item twice. (However the user might get confused if the item is still in the media gallery).

If we do put it in the finally clause, we'd have to disable the delete button while the delete operation is ongoing. I'm just a bit worried that it might be seen as "slow", so if the user wants to delete multiple items they're "stuck" on the dialog screen in between each delete. 

I don't feel strongly about either approaches, WDYT?

spacedmonkey commented on assets/src/edit-story/components/library/panes/media/deleteDialog.js:
```diff
@@ -0,0 +1,108 @@
+        <>
+          <Plain onClick={() => onClose()}>{__('Cancel', 'web-stories')}</Plain>
+          <Plain onClick={() => onDelete()}>
+            {__('Delete', 'web-stories')}
```
Could this be more like 'Yes I am sure'

joannag6 replied:
Hmm, I'm not a fan of that, I'd prefer to keep the "Delete" keyword in there so there's less ambiguity. How about changing the Delete button to a Primary button so it's more obvious?

joannag6 committed:
Extracted out dialog text, changed button types

joannag6 committed:
Fixed unit test to check for correct string

joannag6 committed:
Fixed merge conflict

swissspidy commented on assets/src/edit-story/components/library/panes/media/deleteDialog.js:
```diff
@@ -0,0 +1,122 @@
+        <>
+          <Secondary onClick={onClose}>{__('Cancel', 'web-stories')}</Secondary>
+          <Space />
+          <Primary onClick={onDelete}>{__('Delete', 'web-stories')}</Primary>
```
Let's be consistent with the dialog actions already used in `assets/src/edit-story/components/header/buttons.js` and in #1733.

1. Let's remove the spacer
1. Let's make  the cancel button a `<Plain>` one

joannag6 replied:
Hmm, the mocks seem to have the buttons separated by 8px (https://www.figma.com/file/Cd9iacVknHK62RO4pRwb68/Stories-Editor?node-id=7941%3A18814) 

... and also both Plain buttons. Should I revert them back to Plain buttons or keep the "actionable" button as Primary?

(Same for #1733 😅)

swissspidy replied:
Hmm odd. Let's leave as-is then but create a follow-up issue to make things more consistent.

Opened #2080 to make comparison easier.

joannag6 replied:
Opened #2104 as follow up issue :)

swissspidy reviewed (approved):
LGTM with one nit

joannag6 committed:
Use Plain instead of Secondary to be consistent with #1733
